packages/integration-tests/src/entities/types.ts

```diff
@@ -15,3 +16,34 @@ export const AddressSchema = z.object({
 export const quotes = array(string());
 
 export type IpAddress = string & { __type: "IpAddress" };
+
+export class PasswordValue {
+  static fromEncoded(str: string) {
+    return new PasswordValue(str);
+  }
+
+  static fromPlainText(str: string) {
+    return new PasswordValue(Buffer.from(str, "utf8").toString("base64"));
+  }
+
+  constructor(public readonly encoded: string) {}
+
+  matches(str: string) {
+    return Buffer.from(str, "utf8").toString("base64") === this.encoded;
+  }
+}
+
+export class PasswordValueSerde extends Serde {
```

Comment:
This is great; I really like seeing the implementation & your `PasswordValueSerde` impl--seeing them in action, I'm wondering if we could find a way for the user to provide _only_ `fromEncoded` and `toEncoded` (or equivalents).

Like maybe instead of the user providing a full-fledged `Serde` impl, we could have a `CustomSerde` implementation that looked like:

```
export class CustomSerde extends Serde {
  constructor(...the user's encode & decode...) {}
  dbValue(data: any): any {
    return this.toDb(data[...]);
  }
  mapToDb(value: any): any {
    return this.fromDb(value);
  }
  setOnEntity(data: any, row: any): void {
    data[this.fieldName] = this.fromDb(row[this.columnName]);
  }
}
```

As this would a) reduce the boilerplate that user types have to implement, and b) reduce the API coupling of the current `Serde` internal methods (that per my previous disclaimers are "just okay" imo) to the userland serde implementations.

It seems like we could achieve this with two ways:

1. Have `CustomSerde` be an `abstract class` that provided default impls of the existing `dbValue` / `mapToDb` / `setOnEntity` methods, and just asked the user to implement two new simpler methods, like `toDb` & `fromDb`.

2. Keep `CustomSerde` internal/private, and instead ask the user to implement a new "pair of functions" type. Maybe the internal serde is called like `class CustomSerdeAdapter implements Serde` and we export an interface that is just `interface CustomSerde<DomainType, DbType>` with `toDb(dt: DomainType): DbType` and `fromDb(db: DbType): DomainType` methods.

Wdyt about 1 vs. 2? I think I prefer 2...

Granted, the `CustomSerde` `toDb` / `fromDb` methods would make assumptions that the user's serde is only 1 db column <-> 1 domain field, and in theory we should eventually let users do N db columns <-> 1 domain field, i.e. a `Currency` domain type that used two `currency_amount` and `currency_type` db columns for storing a single domain type...

I'm not sure how we'd grow my current `interface CustomSerde` + `toDb` + `fromDb` suggestion into one that also supported this more complicated use case--but that would probably be a ways out anyway. ...and maybe it'd be as simple as adding a `columns: [..., ...]` field to the JSON config, which if present then we'd ask the user to implement a `MultiColumnSerde` that was not as dead simple as the proposed `CustomSerde`+`toDb`+`fromDb`.

Okay, I think that's enough of an idea of how to support multi-column serde "someday" that I'm pretty comfortable doing the option 2 above for now...

Apologies for asking/brainstorming for another rev, but hopefully this will be worth it. Wdyt?

Reply:
I'm a fan of two too! I like the flexability it provides being just a simple object.

The fact we've got the users serde impl being wrapped inside of our CustomSerdeAdapter also doesn't lock us in too much which is nice.

I have a impl of these changes, but I'm considering one thing - should the user's serde be aware of nullability? I'm personally erring on no. Null is represented as the absence of a value and essentially outside the scope of serde as serde is for mapping values. I think this is really the only fesible way to go for now at least, as this currently how the `type` config works and creates questions around default values etc. But thought it was worth flagging to ensure we're on the same page re semantics.

Reply:
Yep, agreed that we can keep nullability logic encapsulated in our internal serde, and let the user's serde just detail with non-null values.